app/forms/analyses/gwas/form.rb

```diff
@@ -0,0 +1,94 @@
+module Analyses
+  module Gwas
+    class Form < BaseForm
+      include Reform::Form::ActiveModel
+
+      model :analysis
+
+      property :analysis_type
+      property :name
+      property :genotype_data_file_id
+      property :map_data_file_id
+      property :phenotype_data_file_id
+
+      validates :name, presence: true
+      validates :genotype_data_file, presence: true
+      validates :phenotype_data_file, presence: true
+
+      validate do
+        geno = parse_data_file(genotype_data_file, genotype_data_parser)
+        map = parse_data_file(map_data_file, map_data_parser)
+        pheno = parse_data_file(phenotype_data_file, phenotype_data_parser)
+
+        if geno && pheno
+          unless geno.valid?
+            geno.errors.each { |error| errors.add(:genotype_data_file, error) }
+          end
+
+          unless pheno.valid?
+            pheno.errors.each { |error| errors.add(:phenotype_data_file, error) }
+          end
+
+          unless geno.sample_ids.try(:sort) == pheno.sample_ids.try(:sort)
+            errors.add(:base, :geno_pheno_samples_mismatch)
+          end
+        end
+
+        if map_data_file
+          unless map.valid?
+            map.errors.each { |error| errors.add(:map_data_file, error) }
+          end
+        end
+
+        if geno && map
+          unless map.mutation_ids.try(:sort) == geno.mutation_ids.try(:sort)
+            errors.add(:base, :geno_map_mutations_mismatch)
+          end
+        end
+      end
+
+      def genotype_data_file
+        data_file = Analysis::DataFile.gwas_genotype.find_by(id: genotype_data_file_id)
+        AnalysisDataFileDecorator.decorate(data_file) if data_file
+      end
+
+      def map_data_file
+        data_file = Analysis::DataFile.gwas_map.find_by(id: map_data_file_id)
+        AnalysisDataFileDecorator.decorate(data_file) if data_file
+      end
+
+      def phenotype_data_file
+        data_file = Analysis::DataFile.gwas_phenotype.find_by(id: phenotype_data_file_id)
```

Comment:
Interesting question: does an additional where help if you're searching by the id anyway?